### DISCUSSION TITLE
Elm-Review LSP Troubles

### DISCUSSION THREAD

**Discussion by ghost** [helix-editor/helix#8305] [Troubleshooting]
Elm has standard diagnostics (ElmLS diagnostics) and much better diagnostics which the community will encourage you to use called Elm-Review ( https://github.com/jfmengels/elm-review ). Elm Review should be the default, but for whatever reason it's not, so you need to enable it in your editor. This works fine in other editors like VSCode, but not in Helix. After some hours of testing by the user gabbydd in the Helix Matrix server, it was determined that this is a problem on the Helix side (though I am a novice and could not explain what the issue is, so they may need to chime in for that part). If another file is opened prior to the file which has Elm-Review hints, then the Elm-Review hints will be visible, otherwise they will not. It would be nice to see this fixed. The whole troubleshooting session can be found here in the Helix Matrix server: https://matrix.to/#/!zMuVRxoqjyxyjSEBXc:matrix.org/$H6nsD-aGU2ut9g6lS6sO1S2xbzBPuiw_VGWxxbtg2y8?via=matrix.org&via=mozilla.org&via=envs.net

**Comment by gabydd** [comment]
this is quite interesting cause what happens for the ``didOpen`` part is helix sends this after the lsp sends the initialized notification (source: https://github.com/helix-editor/helix/blob/7fa5f341e931f9bdb4058e51ebf77276e9b82242/helix-term/src/application.rs#L681-L713) I think this is probably correct on the helix side but it causes a bit of a problem on the elmLS side if this is the first file you have opened because they initialize the diagnostic provider after the send the initialized notification but there is a bit of a race condition that the ``didOpen`` notification always wins where the ``didOpen`` notification is sent before the diagnostics provider is initialized. You can actually see this behavior in vscode if you leave the Main.elm file open then close and reopen vscode and you wont see the elm-review diagnostics. For the didSave behavior I think this can be fixed on the elmLS side if they change https://github.com/elm-tooling/elm-language-server/blob/44168cf007dbbc6153e660c192b69104ba159413/src/common/capabilityCalculator.ts#L45 to: ``      textDocumentSync: { save: true, change: TextDocumentSyncKind.Incremental, openClose: true },`` ah but as I just found we should be sending the didsave actually: see https://github.com/microsoft/language-server-protocol/issues/288 and https://github.com/neovim/neovim/issues/13989 I can probably make a pr in the next couple days that aligns us to this table
